add testing that a prototypical dandiset metadata validates using converted to json schema

`tools/pubschemata.py` publishes schema as json schema to https://github.com/dandi/schema .  Conversion to json might "mutate" it so a correct according to pydantic's model metadata would not validate using that json schema and `jsonschema` python package (is that what you were using @dchiquit? found no jsonschema import in dandi-api though).

- move `publish_model_schemata` from `tools/pubschemata.py`  into `dandi/metadata.py`
- add a function to validate metadata given provided json schema directory (as produced by `publish_model_schemata`)
- verify that metadata validates using that schema and jsonschema:
   - add tests: in http://github.com/dandi/dandi-cli/blob/HEAD/dandi/tests/test_metadata.py we already have some quite detailed sample cases for dandiset level metadata migration.

@yarikoptic Example, from trying to validate the object in `test_dandimeta_migration`:

```
E   jsonschema.exceptions.ValidationError: None is not of type 'array'
E
E   Failed validating 'type' in schema['properties']['studyTarget']:
E       {'description': 'What the study is related to',
E        'items': {'anyOf': [{'format': 'uri',
E                             'maxLength': 65536,
E                             'minLength': 1,
E                             'type': 'string'},
E                            {'type': 'string'}]},
E        'nskey': 'dandi',
E        'title': 'Study Target',
E        'type': 'array'}
E
E   On instance['studyTarget']:
E       None
```

@yarikoptic I've now run into this error, and I'm not sure what part should be modified to address it:

```
E   jsonschema.exceptions.ValidationError: 'CC-BY-4.0' is not of type 'array'
E       
E   Failed validating 'type' in schema['properties']['license']:
E       {'description': 'License of item.',
E        'items': {'$ref': '#/definitions/LicenseType'},
E        'nskey': 'schema',
E        'title': 'License',
E        'type': 'array'}
E
E   On instance['license']:
E       'CC-BY-4.0'
```

--- a/dandi/metadata.py
+++ b/dandi/metadata.py
@@ -249,11 +249,13 @@ def extract_wasDerivedFrom(metadata):
 
 
 def extract_wasAttributedTo(metadata):
-    return [
-        extract_model(
-            models.Participant, metadata, identifier=metadata.get("subject_id")
-        )
-    ]
+    wat = extract_model(
+        models.Participant, metadata, identifier=metadata.get("subject_id")
+    )
+    if all(v is None for v in wat.dict().values()):
+        return []
+    else:
+        return [wat]
 
 
 def extract_digest(metadata):

--- a/dandi/metadata.py
+++ b/dandi/metadata.py
@@ -1,8 +1,10 @@
 from datetime import datetime
+import json
 import os.path as op
 from pathlib import Path
 import re
 from uuid import uuid4
+import jsonschema
 from . import models
 from .pynwb_utils import (
     _get_pynwb_metadata,
@@ -526,3 +528,16 @@ def publish_model_schemata(releasedir):
     vdir.mkdir(exist_ok=True, parents=True)
     (vdir / "dandiset.json").write_text(models.DandiMeta.schema_json(indent=2))
     (vdir / "asset.json").write_text(models.AssetMeta.schema_json(indent=2))
+    return vdir
+
+
+def validate_dandiset_json(data, schema_dir):
+    with Path(schema_dir, "dandiset.json").open() as fp:
+        schema = json.load(fp)
+    jsonschema.validate(data, schema)
+
+
+def validate_asset_json(data, schema_dir):
+    with Path(schema_dir, "asset.json").open() as fp:
+        schema = json.load(fp)
+    jsonschema.validate(data, schema)

--- a/dandi/metadata.py
+++ b/dandi/metadata.py
@@ -1,5 +1,6 @@
 from datetime import datetime
 import os.path as op
+from pathlib import Path
 import re
 from uuid import uuid4
 from . import models
@@ -517,3 +518,11 @@ def migrate2newschema(meta):
     newmeta = convertv1(meta)
     dandimeta = models.DandiMeta.unvalidated(**newmeta)
     return dandimeta
+
+
+def publish_model_schemata(releasedir):
+    version = models.CommonModel.__fields__["schemaVersion"].default
+    vdir = Path(releasedir, version)
+    vdir.mkdir(exist_ok=True, parents=True)
+    (vdir / "dandiset.json").write_text(models.DandiMeta.schema_json(indent=2))
+    (vdir / "asset.json").write_text(models.AssetMeta.schema_json(indent=2))

--- a/dandi/models.py
+++ b/dandi/models.py
@@ -1,5 +1,5 @@
 from enum import Enum
-from pydantic import BaseModel, Field, AnyUrl, EmailStr, validator
+from pydantic import BaseModel, ByteSize, Field, AnyUrl, EmailStr, validator
 from typing import Dict, List, Union, Optional, Any, Type
 from datetime import date
 from ruamel import yaml
@@ -15,6 +15,8 @@ from .model_types import (
 )
 
 from .utils import name2title
+
+TempOptional = Optional
 
 
 def create_enum(data):
@@ -592,16 +594,16 @@ class DandiMeta(CommonModel):
         min_items=1,
     )
 
-    citation: str = Field(readOnly=True, nskey="schema")
+    citation: TempOptional[str] = Field(readOnly=True, nskey="schema")
 
     # From assets
-    assetsSummary: AssetsSummary = Field(readOnly=True, nskey="dandi")
+    assetsSummary: TempOptional[AssetsSummary] = Field(readOnly=True, nskey="dandi")
 
     # From server (requested by users even for drafts)
-    manifestLocation: List[AnyUrl] = Field(readOnly=True, nskey="dandi")
+    manifestLocation: TempOptional[List[AnyUrl]] = Field(readOnly=True, nskey="dandi")
 
     # On publish
-    version: str = Field(readOnly=True, nskey="schema")
+    version: TempOptional[str] = Field(readOnly=True, nskey="schema")
     doi: Optional[Union[AnyUrl, str]] = Field(
         None, title="DOI", readOnly=True, nskey="dandi"
     )
@@ -634,7 +636,7 @@ class AssetMeta(CommonModel):
         None, description="License of item", nskey="schema"
     )
 
-    contentSize: str = Field(nskey="schema")
+    contentSize: ByteSize = Field(nskey="schema")
     encodingFormat: Union[AnyUrl, str] = Field(
         title="File Encoding Format", nskey="schema"
     )

--- a/dandi/tests/test_metadata.py
+++ b/dandi/tests/test_metadata.py
@@ -1,7 +1,16 @@
 from datetime import datetime, timedelta
+import json
 from dateutil.tz import tzutc
 import pytest
-from ..metadata import metadata2asset, parse_age, timedelta2duration, migrate2newschema
+from ..metadata import (
+    metadata2asset,
+    parse_age,
+    publish_model_schemata,
+    timedelta2duration,
+    migrate2newschema,
+    validate_asset_json,
+    validate_dandiset_json,
+)
 from ..models import (
     AccessRequirements,
     AccessType,
@@ -21,6 +30,11 @@ from ..models import (
     Person,
     Organization,
 )
+
+
+@pytest.fixture(scope="module")
+def schema_dir(tmp_path_factory):
+    return publish_model_schemata(tmp_path_factory.mktemp("schema_dir"))
 
 
 @pytest.mark.parametrize(
@@ -65,8 +79,8 @@ def test_timedelta2duration(td, duration):
     assert timedelta2duration(td) == duration
 
 
-def test_metadata2asset():
-    assert metadata2asset(
+def test_metadata2asset(schema_dir):
+    data = metadata2asset(
         {
             "contentSize": 69105,
             "digest": "783ad2afe455839e5ab2fa659861f58a423fd17f",
@@ -103,7 +117,8 @@ def test_metadata2asset():
                 "Subject",
             ],
         }
-    ) == AssetMeta.unvalidated(
+    )
+    assert data == AssetMeta.unvalidated(
         schemaVersion="1.0.0-rc1",
         identifier="ABC123",
         name=None,
@@ -189,10 +204,14 @@ def test_metadata2asset():
         ],
         contentUrl=None,
     )
+    # We need to convert `data` to a `dict` this way instead of with `.dict()`
+    # so that enums will be converted to strings.
+    data_as_dict = json.loads(data.json(exclude_unset=True, exclude_none=True))
+    validate_asset_json(data_as_dict, schema_dir)
 
 
-def test_metadata2asset_simple1():
-    assert metadata2asset(
+def test_metadata2asset_simple1(schema_dir):
+    data = metadata2asset(
         {
             "contentSize": 69105,
             "digest": "783ad2afe455839e5ab2fa659861f58a423fd17f",
@@ -218,8 +237,10 @@ def test_metadata2asset_simple1():
             "number_of_electrodes": 0,
             "number_of_units": 0,
             "nd_types": [],
+            "tissue_sample_id": "tissue42",
         }
-    ) == AssetMeta.unvalidated(
+    )
+    assert data == AssetMeta.unvalidated(
         schemaVersion="1.0.0-rc1",
         identifier="identifier1",
         name=None,
@@ -257,30 +278,21 @@ def test_metadata2asset_simple1():
         measurementTechnique=None,
         variableMeasured=None,
         wasDerivedFrom=[
-            BioSample.unvalidated(
-                identifier=None, assayType=None, anatomy=None, wasDerivedFrom=None
+            BioSample(
+                identifier="tissue42", assayType=None, anatomy=None, wasDerivedFrom=None
             )
         ],
-        wasAttributedTo=[
-            Participant.unvalidated(
-                identifier=None,
-                source_id=None,
-                strain=None,
-                cellLine=None,
-                vendor=None,
-                age=None,
-                sex=None,
-                genotype=None,
-                species=None,
-                disorder=None,
-            )
-        ],
+        wasAttributedTo=[],
         contentUrl=None,
     )
+    # We need to convert `data` to a `dict` this way instead of with `.dict()`
+    # so that enums will be converted to strings.
+    data_as_dict = json.loads(data.json(exclude_unset=True, exclude_none=True))
+    validate_asset_json(data_as_dict, schema_dir)
 
 
-def test_dandimeta_migration():
-    assert migrate2newschema(
+def test_dandimeta_migration(schema_dir):
+    data = migrate2newschema(
         {
             "dandiset": {
                 "access": {
@@ -539,7 +551,7 @@ def test_dandimeta_migration():
                     "single-neurons",
                 ],
                 "language": "English",
-                "license": "CC-BY-4.0",
+                "license": ["dandi:CCBY40"],
                 "name": (
                     "A NWB-based dataset and processing pipeline of human"
                     " single-neuron activity during a declarative memory task"
@@ -576,7 +588,8 @@ def test_dandimeta_migration():
                 ],
             }
         }
-    ) == DandiMeta.unvalidated(
+    )
+    assert data == DandiMeta.unvalidated(
         access=[
             AccessRequirements(
                 status=AccessType.Open,
@@ -1233,7 +1246,7 @@ def test_dandimeta_migration():
             "open source",
             "single-neurons",
         ],
-        license="CC-BY-4.0",
+        license=["dandi:CCBY40"],
         name=(
             "A NWB-based dataset and processing pipeline of human single-neuron"
             " activity during a declarative memory task"
@@ -1252,3 +1265,7 @@ def test_dandimeta_migration():
         version=None,
         doi=None,
     )
+    # We need to convert `data` to a `dict` this way instead of with `.dict()`
+    # so that enums will be converted to strings.
+    data_as_dict = json.loads(data.json(exclude_unset=True, exclude_none=True))
+    validate_dandiset_json(data_as_dict, schema_dir)

--- a/dandi/tests/test_validate.py
+++ b/dandi/tests/test_validate.py
@@ -7,10 +7,8 @@ def test_validate_simple1(monkeypatch, simple1_nwb):
     errors = validate_file(simple1_nwb)
     assert len(errors) == 1
     assert errors[0] == (
-        "2 validation errors for AssetMeta\n"
+        "1 validation error for AssetMeta\n"
         "wasDerivedFrom -> 0 -> identifier\n"
-        "  none is not an allowed value (type=type_error.none.not_allowed)\n"
-        "wasAttributedTo -> 0 -> identifier\n"
         "  none is not an allowed value (type=type_error.none.not_allowed)"
     )
 

--- a/setup.cfg
+++ b/setup.cfg
@@ -43,6 +43,7 @@ install_requires =
     etelemetry >= 0.2.0
     fasteners
     joblib
+    jsonschema
     pycryptodomex  # for EncryptedKeyring backend in keyrings.alt
     pydantic
     pyout >=0.5, !=0.6.0

--- a/tools/pubschemata.py
+++ b/tools/pubschemata.py
@@ -1,15 +1,5 @@
-from pathlib import Path
 import sys
-from dandi.models import AssetMeta, CommonModel, DandiMeta
-
-
-def publish_model_schemata(releasedir):
-    version = CommonModel.__fields__["schemaVersion"].default
-    vdir = Path(releasedir, version)
-    vdir.mkdir(exist_ok=True, parents=True)
-    (vdir / "dandiset.json").write_text(DandiMeta.schema_json(indent=2))
-    (vdir / "asset.json").write_text(AssetMeta.schema_json(indent=2))
-
+from dandi.metadata import publish_model_schemata
 
 if __name__ == "__main__":
     publish_model_schemata(sys.argv[1])

